### ISSUE TITLE
Target independent javascript client library

### ISSUE THREAD

**Issue by keshavkaul** [OpenVidu/openvidu#210]
Is there going to be any work on extracting out the core js client library features so that it's easy to port to other targets other than the browser, for eg. react native.?

**Comment by micaelgallego** [comment]
In the future we plan to add hooks to allow other WebRTC implementations different then browser. In fact, with ionic iOS we have followed this path.

**Comment by pabloFuente** [comment]
That's correct, we have intention of supporting other frameworks such as React Native or NativeScript, the same way we did with Ionic.

**Comment by keshavkaul** [comment]
Great! I've hacked up a re-write of the browser client core with Webrtc and WebSocket provided as a dependency from the platform. I've only implemented essential features of Joining, Leaving, Publishing, Participant Joined, Participant Left, Participant Published.

I've used Typescript as my language of choice, rxjs peer dependency as my event manager/WebSocket handler.

With this, I am able to plug this into react native and use `react-native-webrtc` package to provide Webrtc dependency along with React Native's own WebSocket support.

Let me know if anyone's interested in this approach for me to invest more time.

**Comment by micaelgallego** [comment]
Really cool!!

Do you plan to contribute this stuff to the project? I mean, we have no experience with react-native, but we want to provide as a supported platform. 

Have you reimplemented all code from scratch or have you refactored openvidu-browser lib? If we have a clear path to follow, we could refactor core components in openvidu-browser lib to be used as dependency for your own code.

Are you interested in contribute to the project?

**Comment by keshavkaul** [comment]
Thanks! Yes, I plan to contribute more to this project. At the moment my organization is generously sponsoring my time on this and I intend to release the codebase under my organization's opensource initiative after having an internal review with my team.

At the moment, I have implemented the core client code with minimal features from scratch using openvidu-browser lib as the reference.

My codebase is platform independent, but as you mentioned, the browser lib can be refactored to add my code as a dependency and similarly we can create a react native lib.

Will be sharing my updates soon!

**Comment by micaelgallego** [comment]
Amazing! Thank you for sharing and make OpenVidu a great platform.

**Comment by pabloFuente** [comment]
Any update on this?

**Comment by pabloFuente** [comment]
React Native is now supported by OpenVidu 2.10.0. It is not part of OpenVidu open source stack though, so developers interested in React Native support should contact OpenVidu team through official [Commercial page](https://openvidu.io/commercial)

**Comment by keshavkaul** [comment]
@pabloFuente @micaelgallego 
Sorry for dropping out of the conversation. I couldn't find the time needed to refine my js client library. Now I'm revisiting this and continuing my work with reconnection feature after network drop. I am facing a `reconnection error` with code `40007`  on trying to connect after a simulated network failure.

I studied the Kurento docs here (https://doc-kurento.readthedocs.io/en/stable/features/kurento_protocol.html#network-issues) and also looked at the browser js code to know the correct jsonrpc request format. I am passing sessionId as the internal rpcSessionId I receive after joining a session.

I'm trying to understand why I'm getting this error. Any help is appreciated.

My Openvidu cloudformation stack is at version 2.5.0
Steps I'm taking is -
1. Open socket connection to Openvidu
2. Setup Ping/Pong Requests
3. Join Session.
4. Few moments after a joining, I disable network.
5. After enabling network, I send a re-connect request.
6. I see the above error as response.

```
sending request: {"jsonrpc":"2.0","id":10,"method":"connect","params":{"sessionId":"ghu5beplkbp2bqf6bds27umg35"}}
received response:  {"id":10,"error":{"code":40007,"message":"reconnection error"},"jsonrpc":"2.0"}
```

**Comment by keshavkaul** [comment]
I see that openvidu is not handling session recovery after network disconnects by looking at this code in Openvidu Server. What must be done to handle session recovery? Any inputs are welcome. I'll try and fix the issue.
https://github.com/OpenVidu/openvidu/blob/bca34e150d78cae431913e19b5d5bbeceb35da7f/openvidu-server/src/main/java/io/openvidu/server/rpc/RpcHandler.java#L600

**Comment by willylee007** [comment]
@keshavkaul Update to latest OpenVidu Server 2.11.0, then try it again. I found OpenVidu 2.10.0 implemented this feature. Maybe you need update some websocket signal message from 2.5.0 to 2.11.0

**Comment by keshavkaul** [comment]
@willylee007 I couldn't find if it's fixed in the release notes for 2.11.0 or 2.10.0. The file I linked in my earlier comment is from the master branch and is the code handling client disconnects in openvidu server where it explicitly removes any sessions of the disconnected client. So my guess is this has not been implemented as of yet.

**Comment by micaelgallego** [comment]
Reconnection is not handled properly in OpenVidu. It is a complex feature to be inplemented in the right way. 

We will take care of it in the future... Until then, try to make some workaround disconnecting and reconnecting.
